crates/primitives/src/receipt.rs

```diff
@@ -44,6 +47,79 @@ impl Receipt {
     }
 }
 
+/// A collection of receipts organized as a two-dimensional vector.
+#[derive(Clone, Debug, PartialEq, Eq, Default)]
+pub struct Receipts {
+    /// A two-dimensional vector of optional `Receipt` instances.
+    pub receipt_vec: Vec<Vec<Option<Receipt>>>,
+}
+
+impl Receipts {
+    /// Create a new `Receipts` instance with an empty vector.
+    pub fn new() -> Self {
+        Self { receipt_vec: vec![] }
+    }
+
+    /// Create a new `Receipts` instance from an existing vector.
+    pub fn from_vec(vec: Vec<Vec<Option<Receipt>>>) -> Self {
+        Self { receipt_vec: vec }
+    }
+
+    /// Returns the length of the `Receipts` vector.
+    pub fn len(&self) -> usize {
+        self.receipt_vec.len()
+    }
+
+    /// Returns `true` if the `Receipts` vector is empty.
+    pub fn is_empty(&self) -> bool {
+        self.receipt_vec.is_empty()
+    }
+
+    /// Push a new vector of receipts into the `Receipts` collection.
+    pub fn push(&mut self, receipts: Vec<Option<Receipt>>) {
+        self.receipt_vec.push(receipts);
+    }
+
+    /// Retrieves a vector of references to receipts for the specified index.
+    pub fn root_slow(&self, index: usize) -> Option<Vec<&Receipt>> {
+        Some(self.receipt_vec[index].iter().map(Option::as_ref).collect())?
+    }
+
+    /// Retrieves gas spent by transactions as a vector of tuples (transaction index, gas used).
+    pub fn gas_spent_by_tx(&self) -> Vec<(u64, u64)> {
+        self.last()
+            .map(|block_r| {
+                block_r
+                    .iter()
+                    .enumerate()
+                    .map(|(id, tx_r)| {
+                        (
+                            id as u64,
+                            tx_r.as_ref()
+                                .expect("receipts have not been pruned")
```

Comment:
this being on `primitives` crate feels weird though.

maybe we change the function signature to `Result<Vec..` and return error if this happens?

Reply:
@rakita @joshieDo I just did a push to update this part returning an error with a `Result`. I would say it's good to solve this problem, then we could possibly transform this error which is currently a string into a specific `BlockValidationError` error.

Reply:
You have added `'str` as a `Error` here, this is not great. You could add a new error item here: https://github.com/paradigmxyz/reth/blob/7a5b6312731dcfd075aeaf38d8f18f087b958d69/crates/primitives/src/prune/part.rs#L23-L29

And at the place where this function is used to have `?` and you wouldn't need to have `gas_spent_by_tx: Result<Vec<(u64, u64)>, &'static str>,` but just `gas_spent_by_tx: Vec<(u64,u64)>,`

Reply:
@rakita Done